shared/src/iosMain/kotlin/com/twilio/passkeys/AuthorizationControllerWrapper.kt

```diff
@@ -0,0 +1,173 @@
+package com.twilio.passkeys
+
+import com.twilio.passkeys.exception.TwilioException
+import com.twilio.passkeys.exception.UNKNOWN_ERROR
+import com.twilio.passkeys.models.AuthenticatePasskeyResponse
+import com.twilio.passkeys.models.CreatePasskeyResponse
+import com.twilio.passkeys.utils.DeviceUtils
+import platform.AuthenticationServices.ASAuthorization
+import platform.AuthenticationServices.ASAuthorizationController
+import platform.AuthenticationServices.ASAuthorizationControllerDelegateProtocol
+import platform.AuthenticationServices.ASAuthorizationPlatformPublicKeyCredentialAssertion
+import platform.AuthenticationServices.ASAuthorizationPlatformPublicKeyCredentialRegistration
+import platform.AuthenticationServices.ASAuthorizationPublicKeyCredentialAttachment
+import platform.Foundation.NSError
+import platform.darwin.NSObject
+
+interface IAuthorizationControllerWrapper {
+  fun createPasskey(
+    authController: ASAuthorizationController,
+    completion: (CreatePasskeyResult) -> Unit,
+  )
+
+  fun authenticatePasskey(
+    authController: ASAuthorizationController,
+    completion: (AuthenticatePasskeyResult) -> Unit,
+  )
+}
+
+/**
+ * Wraps the ASAuthorizationController functionality in order to improve testability & maintainability.
+ *
+ * @property deviceUtils The utility class for device-related operations.
+ */
+class AuthorizationControllerWrapper : IAuthorizationControllerWrapper {
+  private lateinit var authController: ASAuthorizationController
+  private var createPasskeyCompletion: ((CreatePasskeyResult) -> Unit)? = null
+  private var authenticatePasskeyCompletion: ((AuthenticatePasskeyResult) -> Unit)? = null
+  private val deviceUtils: DeviceUtils = DeviceUtils()
+
+  override fun createPasskey(
+    authController: ASAuthorizationController,
+    completion: (CreatePasskeyResult) -> Unit,
+  ) {
+    this.createPasskeyCompletion = completion
+    this.authController = authController
+    this.authController.delegate = createPasskeyDelegate
+    this.authController.performRequests()
+  }
+
+  override fun authenticatePasskey(
+    authController: ASAuthorizationController,
+    completion: (AuthenticatePasskeyResult) -> Unit,
+  ) {
+    this.authenticatePasskeyCompletion = completion
+    this.authController = authController
+    this.authController.delegate = authenticatePasskeyDelegate
+    this.authController.performRequests()
+  }
+
+  private val createPasskeyDelegate =
+    object : NSObject(), ASAuthorizationControllerDelegateProtocol {
+      override fun authorizationController(
+        controller: ASAuthorizationController,
+        didCompleteWithAuthorization: ASAuthorization,
+      ) {
+        val credentialRegistration = didCompleteWithAuthorization.credential as ASAuthorizationPlatformPublicKeyCredentialRegistration
```

Comment:
Is there a way to pass the `credentialRegistration` to TwilioPasskey and applying the logic below in `TwilioPasskey`? `AuthorizationControllerWrapper` could be in charge of completing the interaction / needed code for iOS authorization, but some logic can be kept in `TwilioPasskey`, so it can be tested

Reply:
No unfortunately that `ASAuthorizationPlatformPublicKeyCredentialRegistration` contains a couple of Objects that can't be initialized outside of Apple private logic. So It won't be possible to create that object from Scratch to use it as a mock.